homeassistant/components/prusalink/__init__.py

```diff
@@ -0,0 +1,108 @@
+"""The PrusaLink integration."""
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+from typing import cast
+
+import async_timeout
+from pyprusalink import InvalidAuth, PrusaLink, PrusaLinkError
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
+
+from .const import DOMAIN
+
+PLATFORMS: list[Platform] = [Platform.SENSOR, Platform.CAMERA]
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up PrusaLink from a config entry."""
+    api = PrusaLink(
+        async_get_clientsession(hass),
+        entry.data["host"],
+        entry.data["api_key"],
+    )
+
+    coordinators = {
+        "printer": PrinterUpdateCoordinator(hass, api),
+        "job": JobUpdateCoordinator(hass, api),
+    }
+    for coordinator in coordinators.values():
+        await coordinator.async_config_entry_first_refresh()
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinators
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+class PrusaLinkUpdateCoordinator(DataUpdateCoordinator):
+    """Update coordinator for the printer."""
+
+    config_entry: ConfigEntry
+    api_method: str
+
+    def __init__(self, hass: HomeAssistant, api: PrusaLink) -> None:
+        """Initialize the update coordinator."""
+        self.api = api
+
+        super().__init__(
+            hass, _LOGGER, name=DOMAIN, update_interval=timedelta(seconds=30)
+        )
+
+    async def _async_update_data(self) -> dict:
+        """Update the data."""
+        try:
+            with async_timeout.timeout(5):
+                return cast(dict, await getattr(self.api, self.api_method)())
```

Comment:
I'd recommend implementing a method, instead of a string, that calls a specific api method so we get typed calls.

The library should preferably be extended to use pydantic so each api call response get completely typed.

Reply:
Added TypedDict to the lib now and typed this call.

Reply:
Ok. I'd still recommend extracting the api call to an abstractmethod so we don't need to cast here.

Reply:
Done.